Hazel/src/Hazel/Scene/SceneSerializer.cpp

```diff
@@ -82,70 +82,86 @@ namespace Hazel {
+		out << YAML::Key << "PerspectiveNear" << YAML::Value << camera.GetPerspectiveNearClip();
+		out << YAML::Key << "PerspectiveFar" << YAML::Value << camera.GetPerspectiveFarClip();
+		out << YAML::Key << "OrthographicSize" << YAML::Value << camera.GetOrthographicSize();
+		out << YAML::Key << "OrthographicNear" << YAML::Value << camera.GetOrthographicNearClip();
+		out << YAML::Key << "OrthographicFar" << YAML::Value << camera.GetOrthographicFarClip();
+		out << YAML::EndMap; // Camera
+
+		out << YAML::Key << "Primary" << YAML::Value << cameraComponent.Primary;
+		out << YAML::Key << "FixedAspectRatio" << YAML::Value << cameraComponent.FixedAspectRatio;
+
+		out << YAML::EndMap; // CameraComponent
+	}
 
-		if (entity.HasComponent<CameraComponent>())
-		{
-			out << YAML::Key << "CameraComponent";
-			out << YAML::BeginMap; // CameraComponent
-
-			auto& cameraComponent = entity.GetComponent<CameraComponent>();
-			auto& camera = cameraComponent.Camera;
-
-			out << YAML::Key << "Camera" << YAML::Value;
-			out << YAML::BeginMap; // Camera
-			out << YAML::Key << "ProjectionType" << YAML::Value << (int)camera.GetProjectionType();
-			out << YAML::Key << "PerspectiveFOV" << YAML::Value << camera.GetPerspectiveVerticalFOV();
-			out << YAML::Key << "PerspectiveNear" << YAML::Value << camera.GetPerspectiveNearClip();
-			out << YAML::Key << "PerspectiveFar" << YAML::Value << camera.GetPerspectiveFarClip();
-			out << YAML::Key << "OrthographicSize" << YAML::Value << camera.GetOrthographicSize();
-			out << YAML::Key << "OrthographicNear" << YAML::Value << camera.GetOrthographicNearClip();
-			out << YAML::Key << "OrthographicFar" << YAML::Value << camera.GetOrthographicFarClip();
-			out << YAML::EndMap; // Camera
-
-			out << YAML::Key << "Primary" << YAML::Value << cameraComponent.Primary;
-			out << YAML::Key << "FixedAspectRatio" << YAML::Value << cameraComponent.FixedAspectRatio;
-
-			out << YAML::EndMap; // CameraComponent
-		}
+	template<>
+	static void SerializeComponent(
+		YAML::Emitter& out, const SpriteRendererComponent& spriteRendererComponent)
+	{
+		out << YAML::Key << "SpriteRendererComponent";
+		out << YAML::BeginMap; // SpriteRendererComponent
 
-		if (entity.HasComponent<SpriteRendererComponent>())
-		{
-			out << YAML::Key << "SpriteRendererComponent";
-			out << YAML::BeginMap; // SpriteRendererComponent
+		out << YAML::Key << "Color" << YAML::Value << spriteRendererComponent.Color;
 
-			auto& spriteRendererComponent = entity.GetComponent<SpriteRendererComponent>();
-			out << YAML::Key << "Color" << YAML::Value << spriteRendererComponent.Color;
+		out << YAML::EndMap; // SpriteRendererComponent
+	}
 
-			out << YAML::EndMap; // SpriteRendererComponent
-		}
+	template<typename T>
+	static void TrySerializeComponent(YAML::Emitter& out, const Entity& entity)
+	{
+		if (entity.HasComponent<T>())
+			SerializeComponent(out, entity.GetComponent<T>());
+	}
```

Comment:
First of all, entity can't be const because of the `entity.HasComponent<T>()` and `entity.GetComponent<T>()` are not const either, the HasComponent could be const, but the GetComponent would require extra overloads, which are not worth it imo.. So at least the const modifier on Entity has to go. It creates errors otherwise.

```suggestion
	template<typename T>
	static void TrySerializeComponent(YAML::Emitter& out, Entity& entity)
	{
		if (entity.HasComponent<T>())
			SerializeComponent(out, entity.GetComponent<T>());
	}
```

Next, I am not so font of the "Try do do it" function...
Can I suggest a namechange? Why not call it `SerializeComponent` as well? Function overloading can handle this perfectly...

```suggestion
	template<typename T>
	static void SerializeComponent(YAML::Emitter& out, Entity& entity)
	{
		if (entity.HasComponent<T>())
			SerializeComponent(out, entity.GetComponent<T>());
	}
```

I would then also suggest moving this function to the top, before the other template specifications using the actual components (thus before line 84)

Reply:
I feel the same about that name, especially since it suggests exception handling. How would you feel about `SerializeComponentIfExists`? I'm also okay with your suggestion; if that's what you prefer I'll support that.

Reply:
I think `SerializeComponent` would be the best option, I first opted for `SerializeComponent_s` or something, but I tried especially using the same name if it would recognize the function overloading, it looks consistent in the implementation of SerializeEntity and SerializeScene